[CSS] Support chaining pseudo-elements after ::details-content

mdubet opened the pull request:
#### 65fe8cf1723d8393749fdb5d5e5dd67cc60fd864
<pre>
[CSS] Support chaining pseudo-elements after ::details-content
<a href="https://bugs.webkit.org/show_bug.cgi?id=283446">https://bugs.webkit.org/show_bug.cgi?id=283446</a>
<a href="https://rdar.apple.com/140328987">rdar://140328987</a>

Reviewed by Antti Koivisto.

* LayoutTests/TestExpectations:

Pass new tests

* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isSimpleSelectorValidAfterPseudoElement):
(WebCore::CSSSelectorParser::consumeCompoundSelector):

Allow parsing of pseudo-element after ::details-content.

* Source/WebCore/css/parser/CSSSelectorParser.h:

Store the complete preceding pseudo-element instead of only its type
to be able to check what kind of user agent part it is.

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolvePseudoElement):

When inside a user agent part, we also check the enclosing resolver if necessary for
the boolean firstLine and firstLetter.

Canonical link: <a href="https://commits.webkit.org/300387@main">https://commits.webkit.org/300387@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/921e905103494f138a38e4e54b7fda45b85376c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128989 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e2a73ead-969b-4baf-8861-71b43d076c86) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124276 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42825 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50699 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93030 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2b7338ce-0f9a-4584-a22a-2717cbb7ea0c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125352 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34155 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109599 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73689 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/badf73cf-7045-445e-921b-80dbd8cda3f8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33142 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27756 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72483 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103770 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27966 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131728 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49341 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37550 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101580 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49715 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105819 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101450 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25727 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46827 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24962 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49199 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54942 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48666 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52016 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50348 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -430,10 +432,30 @@ std::optional<ElementUpdate> TreeResolver::resolvePseudoElement(Element& element
+
+    auto shadowTreeEnclosingResolver = [&] -> Resolver* {
+        if (auto* htmlElement = dynamicDowncast<HTMLElement>(element)) {
+            if (htmlElement->isInUserAgentShadowTree())
```
I don't think you need to cast to HTMLElement here?

anttijk commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -430,10 +432,30 @@ std::optional<ElementUpdate> TreeResolver::resolvePseudoElement(Element& element
-    if (pseudoElementIdentifier.pseudoId == PseudoId::FirstLetter && !scope().resolver->usesFirstLetterRules())
-        return { };
+
+    auto shadowTreeEnclosingResolver = [&] -> Resolver* {
```
userAgentShadowTreeEnclosingResolver?

mdubet committed:
[CSS] Support chaining pseudo-elements after ::details-content
https://bugs.webkit.org/show_bug.cgi?id=283446
rdar://140328987

Reviewed by Antti Koivisto.

* LayoutTests/TestExpectations:

Pass new tests

* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isSimpleSelectorValidAfterPseudoElement):
(WebCore::CSSSelectorParser::consumeCompoundSelector):

Allow parsing of pseudo-element after ::details-content.

* Source/WebCore/css/parser/CSSSelectorParser.h:

Store the complete preceding pseudo-element instead of only its type
to be able to check what kind of user agent part it is.

* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):

* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolvePseudoElement):

When inside a user agent part, we also check the enclosing resolver if necessary for
the boolean firstLine and firstLetter.

Canonical link: https://commits.webkit.org/300387@main
